hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java

```diff
@@ -0,0 +1,378 @@
+  private final AtomicLong runCount;
+
+  private final OzoneManager ozoneManager;
+  private final OmSnapshotManager omSnapshotManager;
+  private final SnapshotChainManager chainManager;
+  private final AtomicBoolean suspended;
+  private final OzoneConfiguration conf;
+  private final AtomicLong successRunCount;
+  private final long snapshotDeletionPerTask;
+
+  public SnapshotDeletingService(long interval, long serviceTimeout,
+      OzoneManager ozoneManager) throws IOException {
+    super("SnapshotDeletingService", interval, TimeUnit.MILLISECONDS,
+        SNAPSHOT_DELETING_CORE_POOL_SIZE, serviceTimeout);
+    this.ozoneManager = ozoneManager;
+    this.omSnapshotManager = ozoneManager.getOmSnapshotManager();
+    this.chainManager = ozoneManager.getSnapshotChainManager();
+    this.runCount = new AtomicLong(0);
+    this.successRunCount = new AtomicLong(0);
+    this.suspended = new AtomicBoolean(false);
+    this.conf = ozoneManager.getConfiguration();
+    this.snapshotDeletionPerTask = conf
+        .getLong(SNAPSHOT_DELETING_LIMIT_PER_TASK,
+        SNAPSHOT_DELETING_LIMIT_PER_TASK_DEFAULT);
+  }
+
+  private class SnapshotDeletingTask implements BackgroundTask {
+
+    @Override
+    public BackgroundTaskResult call() throws Exception {
+      if (!shouldRun()) {
+        return BackgroundTaskResult.EmptyTaskResult.newResult();
+      }
+
+      runCount.incrementAndGet();
+
+      Table<String, SnapshotInfo> snapshotInfoTable =
+          ozoneManager.getMetadataManager().getSnapshotInfoTable();
+      try (TableIterator<String, ? extends Table.KeyValue
+          <String, SnapshotInfo>> iterator = snapshotInfoTable.iterator()) {
+        iterator.seekToFirst();
+
+        long snapshotLimit = snapshotDeletionPerTask;
+
+        while (iterator.hasNext() && snapshotLimit > 0) {
+          SnapshotInfo snapInfo = iterator.next().getValue();
+          SnapshotInfo.SnapshotStatus snapshotStatus =
+              snapInfo.getSnapshotStatus();
+
+          // Only Iterate in deleted snapshot
+          if (!snapshotStatus.equals(
+              SnapshotInfo.SnapshotStatus.SNAPSHOT_DELETED)) {
+            continue;
+          }
+
+          OmSnapshot omSnapshot = (OmSnapshot) omSnapshotManager
+              .checkForSnapshot(snapInfo.getVolumeName(),
+                  snapInfo.getBucketName(),
+                  getSnapshotPrefix(snapInfo.getName()));
+
+          Table<String, RepeatedOmKeyInfo> snapshotDeletedTable =
+              omSnapshot.getMetadataManager().getDeletedTable();
+
+          if (snapshotDeletedTable.getEstimatedKeyCount() == 0) {
+            continue;
+          }
+
+          // Get bucketInfo for the snapshot bucket to get bucket layout.
+          String dbBucketKey = ozoneManager.getMetadataManager().getBucketKey(
+              snapInfo.getVolumeName(), snapInfo.getBucketName());
+          OmBucketInfo bucketInfo = ozoneManager.getMetadataManager()
```

Comment:
If bucket is removed in active db, this might be null in this case.

Reply:
@sumitagrawl Theoretically it shouldn't because we prevent buckets deletion when they still have any snapshots. But yes we should handle the case (throw exception).

Reply:
@smengcl This is possible when getting bucketInfo from active db with volume/bucket name from snapshotDB

Reply:
The buckets cannot be deleted as long as there are snapshots associated with them. So even in active db it should have the bucketInfo. But as @smengcl mentioned, we can handle NPE cases better.